src/lib/formats/rpk.cpp

```diff
@@ -0,0 +1,459 @@
+	if (sha1)
+	{
+		if (!hashes.has_value())
+			hashes.emplace();
+		hashes->add_from_string(hash_collection::HASH_SHA1, *sha1);
+	}
+
+	// finally add the socket
+	m_sockets.emplace_back(*this, std::move(id), rpk_socket::socket_type::ROM, std::string(*file), std::move(hashes), std::nullopt);
+	return std::error_condition();
+}
+
+
+//-------------------------------------------------
+//  add_ram_socket
+//-------------------------------------------------
+
+std::error_condition util::rpk_file::add_ram_socket(std::string &&id, const util::xml::data_node &ram_resource_node)
+{
+	// find the length attribute
+	std::string const *const length_string = ram_resource_node.get_attribute_string_ptr("length");
+	if (!length_string)
+		return rpk_reader::error::MISSING_RAM_LENGTH;
+
+	// parse it
+	unsigned int length;
+	char suffix = '\0';
+	sscanf(length_string->c_str(), "%u%c", &length, &suffix);
+	switch (tolower(suffix))
+	{
+	case 'k':
+		// kilobytes
+		length *= 1024;
+		break;
+
+	case 'm':
+		// megabytes
+		length *= 1024 * 1024;
+		break;
+
+	case '\0':
+		break;
+
+	default:  // failed
+		return rpk_reader::error::INVALID_RAM_SPEC;
+	}
+
+	// determine the type of RAM
+	rpk_socket::socket_type type;
+	std::string const *const ram_type = ram_resource_node.get_attribute_string_ptr("type");
+	if (ram_type && *ram_type == "persistent")
+		type = rpk_socket::socket_type::PERSISTENT_RAM;
+	else
+		type = rpk_socket::socket_type::RAM;
+
+	// persistent RAM needs a file name
+	std::string file;
+	if (type == rpk_socket::socket_type::PERSISTENT_RAM)
+	{
+		std::string const *const ram_filename = ram_resource_node.get_attribute_string_ptr("file");
+		if (ram_filename == nullptr)
+			return rpk_reader::error::INVALID_RAM_SPEC; // <ram type='persistent'> must have a 'file' attribute
+		file = *ram_filename;
+	}
+
+	// finally add the socket
+	m_sockets.emplace_back(*this, std::move(id), type, std::move(file), std::nullopt, length);
+	return std::error_condition();
+}
+
+
+/***************************************************************************
+	RPK SOCKET
+***************************************************************************/
+
+//-------------------------------------------------
+//  ctor
+//-------------------------------------------------
+
+util::rpk_socket::rpk_socket(rpk_file &rpk, std::string &&id, socket_type type, std::string &&filename, std::optional<hash_collection> &&hashes, std::optional<std::uint32_t> length)
+	: m_rpk(rpk)
+	, m_id(id)
+	, m_type(type)
+	, m_filename(filename)
+	, m_hashes(std::move(hashes))
+	, m_length(length)
+{
+}
+
+
+//-------------------------------------------------
+//  dtor
+//-------------------------------------------------
+
+util::rpk_socket::~rpk_socket()
+{
+}
+
+
+//-------------------------------------------------
+//  read_file
+//-------------------------------------------------
+
+std::error_condition util::rpk_socket::read_file(std::vector<std::uint8_t> &result) const
```

Comment:
I kinda dislike using `std::vector` here as it forces a particular storage type on the client, which may want to read into a memory region or something.  I guess part of the annoyance is that you *need* a buffer big enough for the entire decompressed content, as the archive file will complain about the buffer being too small and not partially decompress a member.

Reply:
Its worth noting that the preliminary CoCo RPK support follows exactly the pattern you describe, and takes the resulting std::vector and copies it to a fixed size memory region, even mirroring it if necessary.  If you have an idea of what this interface should look like, I'm all ears.